pages/instructorAssessmentInstances/instructorAssessmentInstances.js

```diff
@@ -105,6 +105,14 @@ router.post('/', function(req, res, next) {
                 res.redirect(req.originalUrl);
             });
         });
+    } else if (req.body.__action == 'grade_all') {
+        const assessment_id = res.locals.assessment.id;
+        const close = false;
+        const overrideGradeRate = true;
+        assessment.gradeAllAssessmentInstances(assessment_id, res.locals.authn_user.user_id, close, overrideGradeRate, function(err) {
```

Comment:
This is a potentially long-running function so I think that it should be run under a server jobSequence? My thinking is that this would get the list of assessment instances to grade, but wouldn't lock them. It would then loop over these instances and call the existing `assessment.gradeAssessmentInstance()` function on each of them, which will take care of locking/unlocking each one individually, so that we only hold the lock briefly. Ideally we'd also print output so the instructor would see:
```
Grading assessment instance #1 for mwest@illinois.edu
Grading assessment instance #1 for zilles@illinois.edu
...
```

Reply:
Yes, good idea, I'll take a look at this. With that in mind, about external grading jobs, currently the jobs are started by this process, but the result page is loaded before they complete. Should these jobs be executed as part of the grade all job sequence or in the background? They could still run in parallel, but should the page return a success once they are triggered or once they are finished?

Reply:
Just pushed what seems to be a working version of using a job sequence. My tests seem to work, and I think this is done or close to done, but I never worked with job sequences before, I'd appreciate a review of the implementation. I used `async.eachSeries` inside the job, is that necessary or should I use something simpler? Also, should I list the completed/pending "grade all" jobs somewhere?

Reply:
The use of `eachSeries()` is perfect!